src/util/heap/monotonepageresource.rs

```diff
@@ -157,12 +157,24 @@ impl<VM: VMBinding> PageResource<VM> for MonotonePageResource<VM> {
                 .unwrap()
                 .grow_space(old, bytes, new_chunk);
 
+            let local_side_metadata_per_chunk = self.common().space.unwrap().local_side_metadata_per_chunk();
+
             self.common()
                 .space
                 .unwrap()
                 .common()
                 .mmapper
-                .ensure_mapped(old, required_pages);
+                .ensure_mapped(old, required_pages, &|chunk| {
+                    if !try_map_metadata_space(
+                        chunk,
+                        BYTES_IN_CHUNK,
+                        VM::VMActivePlan::global().global_side_metadata_per_chunk(),
+                        local_side_metadata_per_chunk,
+                    ) {
+                        // TODO(Javad): handle meta space allocation failure
+                        panic!("failed to mmap meta memory");
+                    }
+                });
```

Comment:
Map metadata memory only when the heap chunk is not mmaped before. See fragmented_mapper.rs